src/Remapping/distributed_remapping.jl

```diff
@@ -144,65 +219,482 @@ function Remapper(target_hcoords, target_zcoords, space)
+            vert_interpolation_weights,
+            vert_bounding_indices,
+            fv,
+        )
+    else
+        # Reading values from fv is expensive, so we try to limit the number of reads. We can do
+        # this because multiple target points might be all contained in the same element.
+        prev_v_lo, prev_v_hi, prev_lidx = -1, -1, -1
+        @inbounds for (vindex, (A, B)) in enumerate(vert_interpolation_weights)
+            (v_lo, v_hi) = vert_bounding_indices[vindex]
+
+            for (out_index, h) in enumerate(local_horiz_indices)
+                # If we are no longer in the same element, read the field values again
+                if prev_lidx != h || v_lo != prev_v_lo || v_hi != prev_v_hi
+                    for i in 1:Nq
+                        scratch_field_values[i] = (
+                            A * fv[i, nothing, nothing, v_lo, h] +
+                            B * fv[i, nothing, nothing, v_hi, h]
+                        )
+                    end
+                    prev_v_lo, prev_v_hi, prev_lidx = v_lo, v_hi, h
+                end
+
+                tmp = zero(FT)
+
+                for i in 1:Nq
+                    tmp += I[out_index, i] * scratch_field_values[i]
+                end
+                out[out_index, vindex] = tmp
+            end
+        end
+    end
+end
+
+function _set_interpolated_values!(
+    out::AbstractArray,
+    field::Fields.Field,
+    _scratch_field_values,
+    local_horiz_indices,
+    local_horiz_interpolation_weights,
+    ::Nothing,
+    ::Nothing,
+)
+
+    space = axes(field)
+    FT = Spaces.undertype(space)
+    quad = Spaces.quadrature_style(space)
+    Nq = Quadratures.degrees_of_freedom(quad)
+    fv = Fields.field_values(field)
+
+    hdims = length(local_horiz_interpolation_weights)
+
+    if ClimaComms.device(space) isa ClimaComms.CUDADevice
+        nitems = length(out)
+        nthreads, nblocks = Topologies._configure_threadblock(nitems)
+        @cuda threads = (nthreads) blocks = (nblocks) set_interpolated_values_kernel!(
+            out,
+            local_horiz_interpolation_weights,
+            local_horiz_indices,
+            fv,
+        )
+    else
+        for (out_index, h) in enumerate(local_horiz_indices)
+            out[out_index] = zero(FT)
+            if hdims == 2
+                for j in 1:Nq, i in 1:Nq
+                    out[out_index] +=
+                        local_horiz_interpolation_weights[1][out_index, i] *
+                        local_horiz_interpolation_weights[2][out_index, j] *
+                        fv[i, j, nothing, nothing, h]
+                end
+            elseif hdims == 1
+                for i in 1:Nq
+                    out[out_index] +=
+                        local_horiz_interpolation_weights[1][out_index, i] *
+                        fv[i, nothing, nothing, nothing, h]
+                end
+            else
+                error("Cannot handle $hdims horizontal dimensions")
```

Comment:
Can we check for and throw this error earlier? (so that the GPU path is also protected from this edge case)

Reply:
Done!